### PR TITLE
Add JSON format reference

### LINKED ISSUE #1842: Document JSON application data format
#### Summary

Document JSON application data format.

#### Why do we need this ?

For application developers to know what data and fields they can expect.

#### What is already there? What do you see now?

Documentation to set up MQTT, webhooks and pub/sub integrations.

#### What is missing? What do you want to see?

The format of the upstream and downstream messages.

#### How do you propose to document this?

Since the format is basically the JSONPB TTN flavor of our protos and is shared among various frontends of Application Server, it might be best to have one reference page to which we link from setting up the integrations.

#### Can you do this yourself and submit a Pull Request?

Can review

### PR DESCRIPTION
**Pull request by adriansmares**
#### Summary

Closes #1842 

#### Changes

- Add JSON format reference.

#### Notes for Reviewers

The change in the MQTT doc allows copy pasting (the old one wouldn't parse).

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [ ] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [x] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: doc/content/reference/application-server-data/data-formats/_index.md [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+weight: 10
+---
+
+The streaming data integrations provide multiple formats for the delivery of application uplink messages and downlink queue operations. This section will delve into the common formats used by them.
+
+<!--more-->
+
+## JSON
+
+[comment]: # (The JSON examples given bellow use the javascript tag instead of json since json does not support comments.)
```
Comment (johanstokking) [inline doc/content/reference/application-server-data/data-formats/_index.md]: You can skip this here

File: doc/content/reference/application-server-data/data-formats/_index.md [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+
+The JSON message format is described in the sections bellow.
+
+> Note: Empty fields are omitted from payloads. As such, if a certain field has a value of `""`, `0` or `false` it will not be present in the message.
+
+### JSON Join Messages
+
+The JSON join messages use the following format:
```
Comment (johanstokking) [inline doc/content/reference/application-server-data/data-formats/_index.md]: "Join-accept Messages"

We're in the JSON section already

File: doc/content/reference/application-server-data/data-formats/_index.md [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+
+```javascript
+{
+  "end_device_ids" : {
+    "device_id" : "dev1",                    // Device ID
+    "application_ids" : {
+      "application_id" : "app1"              // Application ID
+    },
+    "dev_eui" : "0004A30B001C0530",          // Device EUI of the end device
+    "join_eui" : "800000000000000C",         // Join EUI of the end device (also known as application EUI in LoRaWAN versions bellow 1.1)
```
Comment (johanstokking) [inline doc/content/reference/application-server-data/data-formats/_index.md]: `below`

Comment (johanstokking) [reply doc/content/reference/application-server-data/data-formats/_index.md]: `DevEUI`, `JoinEUI`, `AppEUI`

File: doc/content/reference/application-server-data/data-formats/_index.md [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+  "end_device_ids" : {
+    "device_id" : "dev1",
+    "application_ids" : {
+      "application_id" : "app1"
+    },
+    "dev_eui" : "0004A30B001C0530",
+    "join_eui" : "800000000000000C",
+    "dev_addr" : "01497ECC"
+  },
+  "correlation_ids" : [ "as:up:01E191YN5F2HMBQCBCVRF731VY", "gs:conn:01E191S6ZEWB630NTR45K5QN4Q", "gs:uplink:01E191YMZ2S7ZRTEGATVE2S3HH", "ns:uplink:01E191YMZ2J64K6FEW5F0WE7TQ", "rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01E191YMZ2KK3TJYG5C4XZP8JK" ],
```
Comment (johanstokking) [inline doc/content/reference/application-server-data/data-formats/_index.md]: Put on multiple lines for readability, it's already in a details section

File: doc/content/reference/application-server-data/data-formats/_index.md [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+    "received_at" : "2020-02-17T07:49:09.736532315Z"
+  }
+}
+```
+
+</details>
+
+### JSON Uplink Messages
+
+The JSON uplink messages use the following format:
```
Comment (johanstokking) [inline doc/content/reference/application-server-data/data-formats/_index.md]: Idem

File: doc/content/reference/application-server-data/data-formats/_index.md [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+    "dev_eui" : "0004A30B001C0530",
+    "join_eui" : "800000000000000C",
+    "dev_addr" : "00BCB929"
+  },
+  "correlation_ids" : [ "as:up:01E0WZGT6Y7657CPFPE5WEYDSQ", "gs:conn:01E0WDEC6T5T4XXBAX7S1VMFKE", "gs:uplink:01E0WZGSZWT07NE5TS2APTV1Z9", "ns:uplink:01E0WZGSZXZXZS8RFAWZX0F2FY", "rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01E0WZGSZXGE1KS577PFBWRJEE" ],
+  "received_at" : "2020-02-12T15:15:46.014773143Z",
+  "uplink_message" : {
+    "session_key_id" : "AXA50tHUGUucuzS/bCGMNw==",
+    "f_cnt" : 1,
+    "frm_payload" : "gkHe",
```
Comment (johanstokking) [inline doc/content/reference/application-server-data/data-formats/_index.md]: Add a version with `decoded_payload`

File: doc/content/reference/application-server-data/data-formats/_index.md [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+    "join_eui" : "800000000000000C",         // Join EUI of the end device (also known as application EUI in LoRaWAN versions bellow 1.1)
+    "dev_addr" : "00BCB929"                  // Device address known by the Network Server
+  },
+  "correlation_ids" : [ "as:up:01...", ... ],// Correlation identifiers of the message
+  "received_at" : "2020-02-12T15:15..."      // ISO 8601 UTC timestamp at which the message has been received by the Application Server
+  "uplink_message" : {
+    "session_key_id": "AXA50...",            // Join Server issued identifier for the session keys used by this uplink
+    "f_cnt": 1,                              // Frame counter
+    "f_port": 1,                             // Frame port
+    "frm_payload": "gkHe",                   // Frame payload
```
Comment (johanstokking) [inline doc/content/reference/application-server-data/data-formats/_index.md]: Idem

Comment (johanstokking) [reply doc/content/reference/application-server-data/data-formats/_index.md]: Also that this is `base64`

File: doc/content/reference/application-server-data/data-formats/_index.md [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
...FhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFh",
```
Comment (johanstokking) [inline doc/content/reference/application-server-data/data-formats/_index.md]: Please no

**Comment by adriansmares** [comment]
Unfortunately I don't think that the proto message snippets are user friendly in their current state:
- Sub message links do not work when embedded in the docs (try to click on the messages - the link is relative to the current page instead of the actual page for the message)
- The formatting in the page is hard to understand
- You have to recursively enter in each object to see its body
- Descriptions are available only for the top level fields
- Examples are always 0 values

**Pull request merged**